pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -0,0 +1,376 @@
+	p.TCPClient = &dns.Client{Net: "tcp", Timeout: ProxyForwardTimeout, SingleInflight: true}
+
+	return p, nil
+}
+
+// AddAllowed adds reStr, a regexp, to the DNS lookups the proxy allows.
+func (p *DNSProxy) AddAllowed(reStr, endpointID string) {
+	log.WithField("name", reStr).Debug("Adding allowed DNS FQDN pattern")
+	reStr = prepareNameMatch(reStr)
+
+	p.Lock()
+	defer p.Unlock()
+	p.allowed.Add(reStr, endpointID)
+}
+
+// RemoveAllowed removes reStr from the DNS lookups the proxy allows. It must
+// match the form in AddAllowed exactly (i.e. this isn't removing by regex, but
+// by direct equivalence).
+func (p *DNSProxy) RemoveAllowed(reStr, endpointID string) {
+	log.WithField("name", reStr).Debug("Removing allowed DNS FQDN pattern")
+	reStr = prepareNameMatch(reStr)
+
+	p.Lock()
+	defer p.Unlock()
+	p.allowed.Remove(reStr, endpointID)
+}
+
+// CheckAllowed checks name against the rules added to the proxy, and only
+// returns true if this endpointID was added (via AddAllowed) previously.
+func (p *DNSProxy) CheckAllowed(name, endpointID string) bool {
+	name = strings.ToLower(name)
+	p.Lock()
+	defer p.Unlock()
+	return p.allowed.LookupContainsValue(name, endpointID)
+}
+
+// ServeDNS handles individual DNS requests forwarded to the proxy, and meets
+// the dns.Handler interface.
+// It will:
+//  - Look up the endpoint that sent the request by IP, via LookupEndpointIDByIP.
+//  - Check that the endpoint ID is in the set of values associated with the
+//  DNS query (lowercased). If not, the request is dropped.
+//  - The allowed request is forwarded to the originally intended DNS server IP
+//  - The response is shared via NotifyOnDNSResponse (this will go to a
+//  fqdn/RuleGen instance).
+//  - Write the response to the endpoint.
+func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, r *dns.Msg) {
+	qname := string(r.Question[0].Name)
+	scopedLog := log.WithFields(logrus.Fields{
+		logfields.DNSName: qname,
+		logfields.IPAddr:  w.RemoteAddr()})
+
+	scopedLog.Debug("Handling DNS query from endpoint")
+
+	endpointIP, _, _, err := splitPortHostProto(w.RemoteAddr())
+	if err != nil {
+		scopedLog.WithError(err).Error("cannot extract endpoint IP from DNS request")
+		return
+	}
+	endpointID, err := p.LookupEndpointIDByIP(endpointIP)
+	if err != nil {
+		scopedLog.WithError(err).Error("cannot extract endpoint ID from DNS request")
+		return
+	}
+	scopedLog = scopedLog.WithField(logfields.EndpointID, endpointID)
+
+	// The allowed check is first because we don't want to use DNS responses that
+	// endpoints are not allowed to see.
+	// Note: The cache doesn't know about the source of the DNS data (yet) and so
+	// it won't enforce any separation between results from different endpoints.
+	// This isn't ideal but we are trusting the DNS responses anyway.
+	if !p.CheckAllowed(qname, endpointID) {
+		scopedLog.Info("Rejecting DNS query from endpoint")
+		return
```

Comment:
That's a good idea. Ideally it would trigger a retry sooner, so it can then fail sooner. I wonder what the error will be (and I suspect it will be different for TCP and UDP).